File: Algorithm/python/programers/다리를지나는트럭.py
# https://school.programmers.co.kr/learn/courses/30/lessons/42583?language=python3
from collections import deque

DUMMY_TRUCK = 0
class Bridge(object):
    def __init__(self, length, weight):
        self.max_weight = weight
        self.current_cnt = 0
        self.current_weight = 0
        self.legnth = length
        self.queue = deque()
        for _ in range(length):
            self.queue.append(DUMMY_TRUCK)

    def pop_truck(self):
        truck = self.queue.pop()
        if truck:
            self.current_cnt -= 1
            self.current_weight -= truck

    def push_truck(self, truck):
        if truck == DUMMY_TRUCK:
            self.queue.appendleft(truck)
            return True

        if self.max_weight >= self.current_weight + truck and self.legnth >= self.current_cnt + 1:
            self.current_cnt += 1
            self.current_weight += truck
            self.queue.appendleft(truck)
            return True
        else:
            return False

    def __len__(self):
        return len(self.queue)

    def is_empty(self):
        return len(self.queue) == 0

def solution(bridge_length, weight, truck_weights):
    bridge = Bridge(bridge_length, weight)
    trucks = deque(w for w in truck_weights)
    answer = 0
    while trucks:
        bridge.pop_truck()
        # 트럭이 다리에 올라갈 수 있다면
        is_pushed = bridge.push_truck(trucks[0])

        if is_pushed:
            trucks.popleft()
        else:
            bridge.push_truck(DUMMY_TRUCK)

        answer += 1
        
    answer += len(bridge)
        
    return answer

File: Algorithm/C++/프로그래머스/다리를지나는트럭.cpp
#include <string>
#include <vector>
#include <deque>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {

    int answer = 0;
    int currentWeight = 0;
    int currentCnt = 0;
    deque<int> q;

    for(int i = 0; i < bridge_length; i ++)
    {
        q.push_back(0);
    }

    for(int i = 0; i < truck_weights.size(); i ++)
    {
        // 마지막 deque에서 트럭을 꺼낸다. 트럭이 있다면 무게와 개수를 줄이고 없다면 무시
        int truck = q.back();
        q.pop_back();
        // 트럭이 존재한다면
        if(truck > 0) {
            currentWeight -= truck;
            currentCnt -= 1;
        }

        // 주어진 조건에 만족하면 트럭을 추가한다.
        int new_truck = truck_weights[i];
        if(weight >= currentWeight + new_truck && bridge_length >= currentCnt + 1) {
            q.push_front(new_truck);
            currentWeight += new_truck;
            currentCnt += 1;
        }
        // 넣을 수 없는 트럭이 없다면 다음 리터레이션에서 트럭을 넣을 수 있게 i를 증감한다.
        else {
            q.push_front(0);
            i -= 1;
        }

        answer += 1;
    }

    while(!q.empty())
    {
        q.pop_back();
        answer += 1;
    }

    
    return answer;
}